Steel sheet check

3d-gussner opened the pull request:
Thanks @jfestrada for the idea in #4569. 

This is a different approach for this request.

We have the https://help.prusa3d.com/materials (which is missing the PP powder coated steel sheet) and we could use it to warn users when they start to print:

- [X] Add new parameter Type to the `M850`  gcode and `HW Setup -> Sheets` to set the type of sheet
  - [X] Smooth
  - [X] Textured
  - [X] Satin
  - [X] NylonPA
  - [X] PP
  - [X] Custom

  - [X] Reuse and add eeprom values needed
    - The Bed and PINDA temp values within the sheets eeprom aren't / have never been used. Replace these Sheet type
previous values are `>10` as the MINtemp for Bed would not allow any lower values so resetting these with `255` or default values `0-6` is not an issue

  - [X] Also add `M862.7` to check the compatible sheets with parameter `P`. In the PrusaSlicer Filament profiles we could add it to the custom G-code
As the materials help article shows compatibility :heavy_check_mark:  and warnings :grey_exclamation:and not to use :x: I would suggest a bit based value
bit 0 = Smooth compatible
bit 1 = Textured compatible
bit 2 = Satin compatible
bit 3 = NylonPA compatible
bit 4 = PP compatible
bit 5 = Custom sheet compatible (like G10 sheets see https://www.youtube.com/watch?v=g0PK4oXbJT8)

- [X] Add an additional Check state `Always` or `Wait` which is the same as `Warn` but doesn't timeout and HAS to be confirmed on LCD
  - [X] With `Always` any `M115` Gcode will trigger the LCD message to be confirmed.

- [X] Of course the default setting for the Sheet type check would be warn (with timeout)

**ToDos:**
  - [x] Bit check `M862.7 P` 
  - [x] And an additional `M862.7` parameter `Warn` with same bit based value
       - Example:
       - PLA : M862.7 P7 W0 PLA is compatible on Smooth, Textured and Satin,  no additional warnings BUT NOT compatible on NylonPA and PP
       - ASA: M862.7 P4 W3 ASA is compatible on Satin, warning to use glue stick on Smooth and Textured and BUT NOT compatible on NylonPA

- [x] In case there is a warning or stop print send also an `//action:notification` to inform host users.

In PrusaConnect we already have some sheet information

- [ ] Another check we could add is in the Preheat menu, so users which use that will be warned before the starting a print or 1st layer calibration.

- [ ] Update Materials help page as PP sheet is missing
- [ ] Optional Add [prusa.io/materials](http://prusa.io/materials) link we could use in the warnings
  - This will need new translations as we are then limited to 2 rows for the message and the 3rd would be used for the link.  
- [ ] Update PrusaSlicer filament profiles for MK2.5/S and MK3/S/+ printers.

- [ ] Review translations for `Check selected steel sheet.`
   - [x] :czech_republic: `Zkontrolujte tiskový plát.` @DRracer 
   - [X] :de: `Überprüfe ausgewähltes Stahlblech.` @3d-gussner
   - [ ] :es: `Verifique la lámina de acero seleccionada.`
   - [x] :fr: `Vérifiez la plaque en acier sélectionnée.` @ErwanAliasr1 @Ilovemyhous
   - [ ] :croatia: `Provjerite odabrani čelični ploca.`
   - [ ] :hungary: `Nezd meg a kiválasztott az acellapot.` @AttilaSVK @Hauzman
   - [x] :it: `Controllare la piastra d'acciaio selezionata.` @wavexx 
   - [x] :netherlands: `Controleer de geselecteerde staalplaat.` @vintagepc @3d-gussner @stelgenhof
   - [ ] :norway: `Sjekk valgt stålplaten.`
   - [ ] :poland: `Sprawdź wybraną płyta stalowa.` @arekm
   - [x] :romania: `Verificați suprafața de print selectată` @leptun @Hauzman
   - [x] :slovakia: `Skontrolujte vybranú platňu.` @ingbrzy @shatter136
   - [ ] :sweden: `Kontrollera vald metallskivan.`


**Post release tasks:**

- [ ] Update doxygen documentation
- [ ] Update RepRap Gcode wiki
- [ ] PrusaSlicer profiles PR for MK2.5/S and MK3/S/+

**Test PR:**
- In PrusaSlicer 2.9.0 -> Printer Tab -> Custom G-code -> Start G-code add below
```
M862.3 P "[printer_model]" ; printer model check
M862.1 P[nozzle_diameter] ; nozzle diameter check
```
following lines
```
;Sheet compatibility see https://help.prusa3d.com/materials

;Smooth   SMO  1
;Textured TXT  2
;Satin    SAT  4
;Nylon    NYL  8
;PP       PP  16
;Custom   CUS 32

; P<value> compatible                 W<value> warning, like "with glue stick"
;      SMO TXT SAT NYL PP  CUS         SMO TXT SAT NYL PP  CUS
;PLA :P 1 + 2 + 4              = 7  | W                        = 0
;PET :P     2 + 4 + 8          = 14 | W 1                      = 1
;ASA :P         4              = 4  | W 1 + 2                  = 3
;ABS :P         4              = 4  | W 1 + 2                  = 3
;PC  :P         4              = 4  | W 1 + 2                  = 3
;CPE :P     2 + 4              = 6  | W 1                      = 1
;PVA :P 1 + 2 + 4              = 7  | W                        = 0
;BVOH:P 1 + 2 + 4              = 7  | W                        = 0
;HIPS:P 1 + 2 + 4              = 7  | W                        = 0
;PP  :P                16      = 16 | W 1                      = 1
;FLEX:P     2                  = 2  | W 1     + 4              = 5
;NGEN:P     2 + 4              = 6  | W 1                      = 1
;PA  :P             8          = 8  | W     2 + 4              = 6
;Wood:P         4              = 4  | W                        = 0
;PVB :P 1     + 4              = 5  | W                        = 0
;PEI :P             8          = 8  | W     2                  = 2

M862.7 P{if filament_type[initial_tool] == "FLEX"}2{elsif filament_type[initial_tool] == "ASA" or filament_type[initial_tool] == "ABS" or filament_type[initial_tool] == "PC" or filament_type[initial_tool] == "WOOD"}4{elsif filament_type[initial_tool] == "PVB"}5{elsif filament_type[initial_tool] == "CPE" or filament_type[initial_tool] == "NGEN"}6{elsif filament_type[initial_tool] == "PLA" or filament_type[initial_tool] == "PVA" or filament_type[initial_tool] == "BVOH" or filament_type[initial_tool] == "HIPS" }7{elsif filament_type[initial_tool] == "PA" or filament_type[initial_tool] == "PEI"}8{elsif filament_type[initial_tool] == "PET" or filament_type[initial_tool] == "PETG"}14{elsif filament_type[initial_tool] == "PP"}16{else}0{endif} W{if filament_type[initial_tool] == "PET" or filament_type[initial_tool] == "PETG" or filament_type[initial_tool] == "CPE" or filament_type[initial_tool] == "PP" or filament_type[initial_tool] == "NGEN"}1{elsif filament_type[initial_tool] == "PEI"}2{elsif filament_type[initial_tool] == "ASA" or filament_type[initial_tool] == "ABS" or filament_type[initial_tool] == "PC"}3{elsif filament_type[initial_tool] == "FLEX"}5{elsif filament_type[initial_tool] == "PA"}6{else}0{endif}; steel sheet check
```
and before
```
M115 U3.14.1 ; tell printer latest fw version
```
  - This will add automatically during slicing the needed `M862.7 P<value> W<value>` to the start gcode.
  - I tried to add the `M862.7` to the PrusaSlicer Filament section, but frankly `Custom G-Code`s are added after the Printer tab Start G-Code. These kind of checks need to be at the beginning, like the Printer Type Nozzle size checks already are!

- In LCD Settings -> HW Setup -> Checks -> Sheet default is
  - [Warn] which will warn the user on LCD screen with a timeout message when
    - the expected/supported sheet isn't selected / placed on the printer
    - the selected sheet has some warnings, like `use glue stick`
- In LCD Settings -> HW Setup -> Steel Sheets -> <Sheet name> you find a new menu `Type`
   - [Smooth | Texture | Satin | NylonPA | PolyPro | Custom]
- Ideally you have at lest two Steel sheets calibrated.

**Test 1**

1. Connect with a serial terminal
  a. ideally enable some logging
2. Select LCD Settings -> HW Setup -> Checks -> Sheet [Warn]
3. Select `Smooth` sheet
4. Send `M862.7 P7 W0` to simulate a PLA print, which can be printed on Smooth, Textured and Satin sheets BUT NOT on NylonPA or PP sheets.
   a. Nothing shown on LCD screen
5. Send `M862.7 P14 W1` to simulate a PET/G print, which can be printed on Textured, Satin and NylonPA sheets BUT NOT on PP sheets. With a glue stick applied the usage of the Smooth sheet is possible and will warn the user.
  a. LCD should show for ~30 seconds
```
Check selected steel
sheet.

>Cont.        Cancel
``` 
6. Select LCD Settings -> HW Setup -> Checks -> Sheet [Strict]
7. Repeat Steps 3 to 5
   a. The LCD screen of step 5 shows and NEEDS to be confirmed on LCD
```
Check selected steel
sheet.

                   v
``` 
8. Select LCD Settings -> HW Setup -> Checks -> Sheet [Always]
9. Repeat Steps 3 to 5
 a. The LCD message is always shown, but it times out and will continue

**Test 2**
1. Connect with a serial terminal
  a. ideally enable some logging
2. Select LCD Settings -> HW Setup -> Checks -> Sheet [Warn]
3. Select `Smooth` sheet
4. Slice a simple short PLA print with the new/modified PrusaSlicer profile
5. Start this PLA print
6. As the Smooth sheet is compatible with PLA no warnings are shown.
7. Slice a simple short PETG print with the new/modified PrusaSlicer profile
8. As the Smooth sheet should be used with glue stick while printing PET/G
  a. LCD should show for ~30 seconds
```
Check selected steel
sheet.

>Cont.        Cancel
``` 
9. Select LCD Settings -> HW Setup -> Checks -> Sheet [Strict]
10. Repeat step 3 to 8 and you should get same results as in Test 1
11.  Select LCD Settings -> HW Setup -> Checks -> Sheet [Always]
12. Repeat step 3 to 8 and you should get same results as in Test 1 with one addition
  b. To also check "old" gcode without `M862.7`, the check is also executed when `M115` is send during a print and shows following screen which DOESN'T timeout and NEEDS user interaction!
```
Check selected steel
sheet.

 Cont.       >Cancel
```
   i. To not delay host connections the `M115` doesn't execute the check when the printer is inactive.
   ii. Frankly at this moment new gcode will show the LCD message twice when `Always` is selected.

3d-gussner committed:
Sheet types init

Fix rebase issues

3d-gussner committed:
Update messsage and translations

ingbrzy commented on lang/po/Firmware_sk.po:
```diff
@@ -2585,6 +2585,12 @@ msgstr "Nie je zavedený žiaden filament."
+#: ../../Firmware/Marlin_main.cpp:5844 ../../Firmware/messages.cpp:163
+#: ../../Firmware/util.cpp:454 ../../Firmware/util.cpp:455
+msgid "Check selected steel sheet."
+msgstr "Skontrolujte vybraný platňa."
```
should be "Skontrolujte vybranú platňu"

3d-gussner replied:
Thanks, will wait for other reviews and commit later

Hauzman commented:
@3d-gussner RO looks good from my perspective

stelgenhof commented:
@3d-gussner NL looks good to me

Ilovemyhous commented:
French gets a green light from my side. 👍

AttilaSVK commented:
Hungarian should be "Ellenőrizd a kiválasztott az tárgyasztal lapot"

3d-gussner commented:
> Hungarian should be "Ellenőrizd a kiválasztott az tárgyasztal lapot"

Thanks for the suggestion. I searched for existing terms and words and for steel sheet I found these:

"Please place steel sheet on heatbed."
"Kerlek, helyzed az acellapot a targyasztalra."

"Please remove steel sheet from heatbed."
"Kerlek, tavolisd el az acellapot az asztalrol."

Can you please also review the rest of the translations (of course so they do not exceed the limits) for "check"?

Now I see "Ellenorizd" and "Nezd meg a" (look at ???) are used for "Check"

Also you can use the diacritics. In this case "Ellenorizd" should be "Ellenőrizd"  or not?

leptun commented:
@3d-gussner For ro I would do `Verificați suprafața de print selectată.` for consistency with the steel sheet naming in the other translations. Still 2 rows so it fits

3d-gussner commented:
> @3d-gussner For ro I would do Verificați suprafața de print selectată. for consistency with the steel sheet naming in the other translations. Still 2 rows so it fits

Thanks

I will wait for potential other changes before committing.

ErwanAliasr1 commented:
Totally agree, thx !

Le mer. 8 janv. 2025 à 17:07, Ilovemyhouse ***@***.***> a
écrit :

> French gets a green light from my side. 👍
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/prusa3d/Prusa-Firmware/pull/4832#issuecomment-2578053791>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCUIHLKSLSE3EYHEMDQ7P32JVEL3AVCNFSM6AAAAABU2EFJVKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNZYGA2TGNZZGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

AttilaSVK commented:
> > Hungarian should be "Ellenőrizd a kiválasztott az tárgyasztal lapot"
> 
> Thanks for the suggestion. I searched for existing terms and words and for steel sheet I found these:
> 
> "Please place steel sheet on heatbed." "Kerlek, helyzed az acellapot a targyasztalra."
> 
> "Please remove steel sheet from heatbed." "Kerlek, tavolisd el az acellapot az asztalrol."
> 
> Can you please also review the rest of the translations (of course so they do not exceed the limits) for "check"?
> 
> Now I see "Ellenorizd" and "Nezd meg a" (look at ???) are used for "Check"
> 
> Also you can use the diacritics. In this case "Ellenorizd" should be "Ellenőrizd" or not?

"Kérlek, helyzed a lapot a tárgyasztalra." - the literal translation is please place sheet on the object table. In Hungarian we don't really say that it's from steel. (kind of like in the Czech original, if I recall correctly)

However, just for consistency sake:
"Kérlek, távolítsd el a lapot a tárgyasztalról." - plase remove the sheet from the object table

Yes, Nézd meg literally means Look at, but it can mean Check as well. I might have used it instead of Ellenőrizd to save two characters. I'm thinking of going through the entire Hungarian translation in the near future to make it more consistent (and to add diacritics)

As for the diacritics, Ellenőrizd is correct.

leptun commented:
> I'm thinking of going through the entire Hungarian translation in the near future to make it more consistent (and to add diacritics)

Please tag me when you start working on that as we may have to make a few changes to the diacritics scripts (enable/disable individual characters from the font based on usage).

3d-gussner committed:
Add Sheet bit check
Add Sheet warning bit check
Improve `M115` steel check when `Always` is selected for host connection
Add some debug output

3d-gussner commented:
@AttilaSVK Thanks for the response. Please also tag @leptun when you are done, so we can try to get all used Hungarian diacritics shown correctly.

3d-gussner committed:
Fix EEPROM_CHECK_SHEET_TYPE reset to `Warn`
after boot.

3d-gussner committed:
Update translations

3d-gussner committed:
Send //action:notifiactions only when M79 is active

3d-gussner committed:
Send host notifications when M862.x triggers

3d-gussner committed:
Update for host notification
During test in German the notification sent to the host used some unkown diacrtics.
We try to keep host releated messages in English only.

3d-gussner committed:
Fix char sheet length

3d-gussner committed:
Fix reset sheet type value

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -365,7 +365,7 @@ void lcdui_print_percent_done(void)
 		if ((nextSheet >= 0) && (sheetNR != nextSheet))
 		{
-			char sheet[sizeof(Sheet::name)];
+			char sheet[sizeof(Sheet::name)+1UL];
```
Idea: Can we use `SheetName` struct? That is, replace `char sheet[sizeof(Sheet::name)+1UL];` with `SheetName sheet;`? That would make the initialization more consistent and avoid off by one error.

Then instead of:

```c
eeprom_read_block(sheet, EEPROM_Sheets_base->s[sheetNR].name, sizeof(Sheet::name));
sheet[sizeof(Sheet::name)] = '\0';
```

You would use:

```c
eeprom_read_block(sheet.c, EEPROM_Sheets_base->s[sheetNR].name, sizeof(sheets.c));
sheet.c[sizeof(sheets.c)] = '\0';
```

3d-gussner replied:
Changing it to
```
			SheetName sheet;
            eeprom_read_block(sheet.c, EEPROM_Sheets_base->s[sheetNR].name, sizeof(sheet.c));
            sheet.c[sizeof(sheet.c)] = '\0';
```
results in more ressource usage: from `248796` to `248820` bytes

gudnimg commented on Firmware/Marlin_main.cpp:
```diff
@@ -7262,7 +7262,7 @@ void process_commands()
 	}
   // Reset Sheet type if out of range
-  if (iType > STEEL_SHEET_TYPES -1) eeprom_update_byte_notify(&EEPROM_Sheets_base->s[iSel].type, 0);
+  if (iType > 32) eeprom_update_byte_notify(&EEPROM_Sheets_base->s[iSel].type, 0);
```
Where does 32 come from? Can we define a constant for it somewhere?

3d-gussner replied:
We have at this moment 5 Sheet types and previous values above 32 should be invalid.
